# Football Player Project

In [1]:
# Import relevant libraries.
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer

In [2]:
# Load the players_combined.csv dataset.
df = pd.read_csv('players_combined.csv')

# Set display options to show all columns.
pd.set_option('display.max_columns', None)

# View the df DataFrame shape and head.
print("The DataFrame shape:", df.shape)
df.head(20)

The DataFrame shape: (18136, 56)


id           name  age  nationality     club position height  weight  \
0   185149    J. Koprivec   29     Slovenia  No Club       GK    6'1  179lbs   
1   189024    M. Banguera   32      Ecuador  No Club       GK    6'2  198lbs   
2   201572      M. Korhut   29      Hungary  No Club       LB   5'11  168lbs   
3   205211      M. Borjan   30       Canada  No Club       GK    6'5  185lbs   
4   205868     R. Kirilov   26     Bulgaria  No Club       LW    5'7  132lbs   
5   207439     L. Paredes   24    Argentina  No Club       CM   5'11  165lbs   
6   208485       L. Ralte   25        India  No Club       LW    5'8  157lbs   
7   216477     T. Nedelev   25     Bulgaria  No Club       RM    5'9  154lbs   
8   217225  J. Lalpekhlua   27        India  No Club       RS    5'9  165lbs   
9   236457  D. Giannoulis   22       Greece  No Club       LB   5'10  143lbs   
10  238205  A. Edathodika   31        India  No Club      LCB    6'0  172lbs   
11  229579     A. Accardi   22        Italy  Palermo       CB   5'11  165lbs   
12  245665    Y. Ouattara   34  Ivory Coast  No Club       RW    5'9  157lbs   
13  233524       U. Singh   22        India  No Club       RM   5'11  165lbs   
14  246065      L. Haquin   20      Bolivia  No Club      LCB   5'11  168lbs   
15  211339     R. Cardozo   28      Bolivia  No Club       LM    5'7  152lbs   
16  219740        Kahraba   24        Egypt  No Club       LM    6'0  168lbs   
17  223052     J. Narváez   27      Ecuador  No Club       RB    6'0  159lbs   
18  163856     C. Dagnall   32      England     Bury       ST    5'8  165lbs   
19  181795  F. Kudryashov   31       Russia  No Club      LCB   5'11  179lbs   

   preferred_foot       work_rate body_type      joined loaned_from  \
0           Right  Medium/ Medium    Normal         NaN         NaN   
1           Right  Medium/ Medium    Stocky         NaN         NaN   
2            Left    High/ Medium    Normal         NaN         NaN   
3           Right  Medium/ Medium    Normal         NaN         NaN   
4           Right     Medium/ Low      Lean         NaN         NaN   
5           Right  Medium/ Medium    Normal         NaN         NaN   
6            Left       High/ Low      Lean         NaN         NaN   
7            Left  Medium/ Medium    Normal         NaN         NaN   
8           Right  Medium/ Medium    Stocky         NaN         NaN   
9            Left    High/ Medium      Lean         NaN         NaN   
10           Left  Medium/ Medium    Normal         NaN         NaN   
11          Right       Low/ High    Normal  2015-07-07         NaN   
12           Left    High/ Medium    Normal         NaN         NaN   
13          Right       High/ Low    Normal         NaN         NaN   
14          Right    High/ Medium      Lean         NaN         NaN   
15           Left  Medium/ Medium    Normal         NaN         NaN   
16          Right    High/ Medium    Normal         NaN         NaN   
17          Right  Medium/ Medium      Lean         NaN         NaN   
18          Right      High/ High    Normal  2018-07-01         NaN   
19           Left  Medium/ Medium    Normal         NaN         NaN   

    contract_valid_until     value  release_clause    wage  overall  \
0                    NaN       NaN             NaN     NaN       68   
1                    NaN       NaN             NaN     NaN       73   
2                    NaN       NaN             NaN     NaN       71   
3                    NaN       NaN             NaN     NaN       73   
4                    NaN       NaN             NaN     NaN       65   
5                    NaN       NaN             NaN     NaN       80   
6                    NaN       NaN             NaN     NaN       61   
7                    NaN       NaN             NaN     NaN       68   
8                    NaN       NaN             NaN     NaN       63   
9                    NaN       NaN             NaN     NaN       65   
10                   NaN       NaN             NaN    

In [3]:
# View the metadata of the df DataFrame.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18136 entries, 0 to 18135
Data columns (total 56 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        18136 non-null  int64  
 1   name                      18136 non-null  object 
 2   age                       18136 non-null  int64  
 3   nationality               18136 non-null  object 
 4   club                      18136 non-null  object 
 5   position                  18136 non-null  object 
 6   height                    18136 non-null  object 
 7   weight                    18136 non-null  object 
 8   preferred_foot            18136 non-null  object 
 9   work_rate                 18136 non-null  object 
 10  body_type                 18136 non-null  object 
 11  joined                    16643 non-null  object 
 12  loaned_from               1264 non-null   object 
 13  contract_valid_until      16643 non-null  float64
 14  value 

As you can see above, importing the dataset into Python has removed the 'N/A' values from the 'loaned_from' column. This needs to be corrected once the dataset is imported into Power BI.

In [4]:
# Handle missing values and categorical features.
X = df[['age', 'overall', 'potential', 'position']]  # Features.
y = df['value']  # Target.

# Split the dataset into rows with known and missing 'value'.
train_data = df[df['value'].notnull()]  # Rows with known values.
X_train = train_data[['age', 'overall', 'potential', 'position']]
y_train = train_data['value']

# Data with missing values in 'value' column for prediction.
missing_data = df[df['value'].isnull()]
X_missing = missing_data[['age', 'overall', 'potential', 'position']]

# Define the preprocessing pipeline.
preprocessor = ColumnTransformer(transformers=[
        ('num', SimpleImputer(strategy='mean'), ['age', 'overall', 'potential']),  # Impute missing numeric values.
        ('cat', OneHotEncoder(handle_unknown='ignore', drop='first'), ['position'])])  # OneHotEncoder for 'position'.

In [5]:
# Create the Random Forest model pipeline.
model_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', RandomForestRegressor(random_state=42))])

# Train the Random Forest model on known values.
model_pipeline.fit(X_train, y_train)

# Predict on the rows with missing values.
predictions = model_pipeline.predict(X_missing)

# Evaluate the Random Forest model on training data.
y_train_pred = model_pipeline.predict(X_train)

# Calculate metrics for Random Forest model.
mae = mean_absolute_error(y_train, y_train_pred)
mse = mean_squared_error(y_train, y_train_pred)
rmse = mean_squared_error(y_train, y_train_pred, squared=False)
r2 = r2_score(y_train, y_train_pred)

# Output evaluation metrics for Random Forest model.
print(f"Random Forest - MAE: {mae}, MSE: {mse}, RMSE: {rmse}, R2: {r2}")

Random Forest - MAE: 55837.91158196265, MSE: 101145115947.5019, RMSE: 318033.1994422939, R2: 0.9968124441860214


The Random Forest model demonstrates strong predictive performance with an R² value of 0.9968, indicating that 99.68% of the variance in the target variable is explained by the model. The MAE of 55,837.91 and RMSE of 318,033.20 reflect the average and root mean square prediction errors, respectively, with the RMSE being more sensitive to large deviations. While the MSE of 101,145,115,947.50 is high due to squaring errors, the overall metrics suggest the model fits the data very well, though careful attention is needed to assess real-world relevance and overfitting risks.

In [6]:
# Create the Gradient Boosting model pipeline.
gb_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),  # Reuse the existing preprocessor.
    ('model', GradientBoostingRegressor(random_state=42))])

# Train the Gradient Boosting model on known values.
gb_pipeline.fit(X_train, y_train)

# Predict on rows with missing values using Gradient Boosting.
gb_predictions = gb_pipeline.predict(X_missing)

# Evaluate the Gradient Boosting model on training data.
y_train_pred_gb = gb_pipeline.predict(X_train)

# Calculate metrics for Gradient Boosting model.
mae_gb = mean_absolute_error(y_train, y_train_pred_gb)
mse_gb = mean_squared_error(y_train, y_train_pred_gb)
rmse_gb = mean_squared_error(y_train, y_train_pred_gb, squared=False)
r2_gb = r2_score(y_train, y_train_pred_gb)

# Output evaluation metrics for Gradient Boosting model.
print(f"Gradient Boosting - MAE: {mae_gb}, MSE: {mse_gb}, RMSE: {rmse_gb}, R2: {r2_gb}")

Gradient Boosting - MAE: 209196.40946721667, MSE: 231934770755.14227, RMSE: 481596.06596726086, R2: 0.9926906502596916


The Gradient Boosting model shows strong predictive performance with an R² value of 0.9927, meaning it explains 99.27% of the variance in the target variable. The MAE of 209,196.41 indicates the average error, while the RMSE of 481,596.07 highlights larger prediction deviations. The MSE of 231,934,770,755.14 is higher due to squared error calculations, reflecting the sensitivity to outliers. While the model performs well, its errors are notably larger compared to the Random Forest model, suggesting potential room for improvement in capturing complex patterns.

In [7]:
# Detect and remove outliers using IQR.
Q1 = X_train[['age', 'overall', 'potential']].quantile(0.25)
Q3 = X_train[['age', 'overall', 'potential']].quantile(0.75)
IQR = Q3 - Q1

# Filter the data to remove outliers.
outlier_condition = ((X_train[['age', 'overall', 'potential']] < (Q1 - 1.5 * IQR)) |
                     (X_train[['age', 'overall', 'potential']] > (Q3 + 1.5 * IQR))).any(axis=1)

# Remove outliers from X_train and y_train.
X_train_no_outliers = X_train[~outlier_condition]
y_train_no_outliers = y_train[~outlier_condition]

# Create the Random Forest model pipeline without outliers.
model_pipeline_no_outliers = Pipeline(steps=[
    ('preprocessor', preprocessor), # Reuse the existing preprocessor.
    ('model', RandomForestRegressor(random_state=42))])

# Train the Random Forest model on data without outliers.
model_pipeline_no_outliers.fit(X_train_no_outliers, y_train_no_outliers)

# Predict on the rows with missing values.
predictions_no_outliers = model_pipeline_no_outliers.predict(X_missing)

# Evaluate the Random Forest model on training data without outliers.
y_train_pred_no_outliers = model_pipeline_no_outliers.predict(X_train_no_outliers)

# Calculate metrics for the Random Forest model without outliers.
mae_no_outliers = mean_absolute_error(y_train_no_outliers, y_train_pred_no_outliers)
mse_no_outliers = mean_squared_error(y_train_no_outliers, y_train_pred_no_outliers)
rmse_no_outliers = mean_squared_error(y_train_no_outliers, y_train_pred_no_outliers, squared=False)
r2_no_outliers = r2_score(y_train_no_outliers, y_train_pred_no_outliers)

# Output evaluation metrics for the Random Forest model without outliers.
print(f"Random Forest without Outliers - MAE: {mae_no_outliers}, MSE: {mse_no_outliers}, "
      f"RMSE: {rmse_no_outliers}, R2: {r2_no_outliers}")

Random Forest without Outliers - MAE: 38462.46172308438, MSE: 19519516139.496857, RMSE: 139712.26195111457, R2: 0.9985660397294006


The Random Forest model without outliers achieves exceptional performance, with an R² value of 0.9986, explaining 99.86% of the variance in the target variable. The MAE of 38,462.46 and RMSE of 139,712.26 indicate significantly lower prediction errors compared to the Random Forest model that included outliers, highlighting improved accuracy. The MSE of 19,519,516,139.50 is also markedly reduced, reflecting the model's ability to better handle the adjusted data distribution. These results demonstrate that removing outliers substantially enhances the model's predictive precision and reliability.

In [8]:
# Predict the missing values using the Random Forest model without outliers.
# Round to 0 decimal place to keep consistent with the other values in the 'value' column.
df.loc[df['value'].isnull(), 'value'] = model_pipeline_no_outliers.predict(X_missing).round(0)

# Check if the 'value' column has any missing values.
missing_values = df['value'].isnull().sum()

# Output the result.
print(f"Number of missing values in 'value' column: {missing_values}")

Number of missing values in 'value' column: 0


In [9]:
# View the metadata of the df DataFrame.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18136 entries, 0 to 18135
Data columns (total 56 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        18136 non-null  int64  
 1   name                      18136 non-null  object 
 2   age                       18136 non-null  int64  
 3   nationality               18136 non-null  object 
 4   club                      18136 non-null  object 
 5   position                  18136 non-null  object 
 6   height                    18136 non-null  object 
 7   weight                    18136 non-null  object 
 8   preferred_foot            18136 non-null  object 
 9   work_rate                 18136 non-null  object 
 10  body_type                 18136 non-null  object 
 11  joined                    16643 non-null  object 
 12  loaned_from               1264 non-null   object 
 13  contract_valid_until      16643 non-null  float64
 14  value 

In [10]:
# Issues existed when importing dataset into Power BI.
# Fill missing values in the specified columns with pd.NA,
# and convert the data types to 'Int64' for handling nullable integers.
df['contract_valid_until'] = df['contract_valid_until'].fillna(pd.NA).astype('Int64')
df['value'] = df['value'].fillna(pd.NA).astype('Int64')
df['release_clause'] = df['release_clause'].fillna(pd.NA).astype('Int64')
df['wage'] = df['wage'].fillna(pd.NA).astype('Int64')

# View the df DataFrame head.
df.head(20)

id           name  age  nationality     club position height  weight  \
0   185149    J. Koprivec   29     Slovenia  No Club       GK    6'1  179lbs   
1   189024    M. Banguera   32      Ecuador  No Club       GK    6'2  198lbs   
2   201572      M. Korhut   29      Hungary  No Club       LB   5'11  168lbs   
3   205211      M. Borjan   30       Canada  No Club       GK    6'5  185lbs   
4   205868     R. Kirilov   26     Bulgaria  No Club       LW    5'7  132lbs   
5   207439     L. Paredes   24    Argentina  No Club       CM   5'11  165lbs   
6   208485       L. Ralte   25        India  No Club       LW    5'8  157lbs   
7   216477     T. Nedelev   25     Bulgaria  No Club       RM    5'9  154lbs   
8   217225  J. Lalpekhlua   27        India  No Club       RS    5'9  165lbs   
9   236457  D. Giannoulis   22       Greece  No Club       LB   5'10  143lbs   
10  238205  A. Edathodika   31        India  No Club      LCB    6'0  172lbs   
11  229579     A. Accardi   22        Italy  Palermo       CB   5'11  165lbs   
12  245665    Y. Ouattara   34  Ivory Coast  No Club       RW    5'9  157lbs   
13  233524       U. Singh   22        India  No Club       RM   5'11  165lbs   
14  246065      L. Haquin   20      Bolivia  No Club      LCB   5'11  168lbs   
15  211339     R. Cardozo   28      Bolivia  No Club       LM    5'7  152lbs   
16  219740        Kahraba   24        Egypt  No Club       LM    6'0  168lbs   
17  223052     J. Narváez   27      Ecuador  No Club       RB    6'0  159lbs   
18  163856     C. Dagnall   32      England     Bury       ST    5'8  165lbs   
19  181795  F. Kudryashov   31       Russia  No Club      LCB   5'11  179lbs   

   preferred_foot       work_rate body_type      joined loaned_from  \
0           Right  Medium/ Medium    Normal         NaN         NaN   
1           Right  Medium/ Medium    Stocky         NaN         NaN   
2            Left    High/ Medium    Normal         NaN         NaN   
3           Right  Medium/ Medium    Normal         NaN         NaN   
4           Right     Medium/ Low      Lean         NaN         NaN   
5           Right  Medium/ Medium    Normal         NaN         NaN   
6            Left       High/ Low      Lean         NaN         NaN   
7            Left  Medium/ Medium    Normal         NaN         NaN   
8           Right  Medium/ Medium    Stocky         NaN         NaN   
9            Left    High/ Medium      Lean         NaN         NaN   
10           Left  Medium/ Medium    Normal         NaN         NaN   
11          Right       Low/ High    Normal  2015-07-07         NaN   
12           Left    High/ Medium    Normal         NaN         NaN   
13          Right       High/ Low    Normal         NaN         NaN   
14          Right    High/ Medium      Lean         NaN         NaN   
15           Left  Medium/ Medium    Normal         NaN         NaN   
16          Right    High/ Medium    Normal         NaN         NaN   
17          Right  Medium/ Medium      Lean         NaN         NaN   
18          Right      High/ High    Normal  2018-07-01         NaN   
19           Left  Medium/ Medium    Normal         NaN         NaN   

    contract_valid_until     value  release_clause  wage  overall  potential  \
0                   <NA>    600250            <NA>  <NA>       68         68   
1                   <NA>   2348000            <NA>  <NA>       73         73   
2                   <NA>   1815000            <NA>  <NA>       71         71   
3                   <NA>   2600000            <NA>  <NA>       73         73   
4                   <NA>    621667            <NA>  <NA>       65         66   
5                   <NA>  19510000            <NA>  <NA>       80         85   
6                   <NA>    298250            <NA>  <NA>       61         62   
7                   <NA>   1157750            <NA>  <NA>       68         73   
8                   <NA>    422000            <NA>  <NA>       63         64   
9                   <NA>    680000 

In [11]:
# View the metadata of the df DataFrame.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18136 entries, 0 to 18135
Data columns (total 56 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   id                        18136 non-null  int64 
 1   name                      18136 non-null  object
 2   age                       18136 non-null  int64 
 3   nationality               18136 non-null  object
 4   club                      18136 non-null  object
 5   position                  18136 non-null  object
 6   height                    18136 non-null  object
 7   weight                    18136 non-null  object
 8   preferred_foot            18136 non-null  object
 9   work_rate                 18136 non-null  object
 10  body_type                 18136 non-null  object
 11  joined                    16643 non-null  object
 12  loaned_from               1264 non-null   object
 13  contract_valid_until      16643 non-null  Int64 
 14  value                 

In [12]:
# Export the updated df DataFrame as players_combined_updated.csv.
df.to_csv('players_combined_updated.csv', index=False)